core/src/test/scala/org/bykn/bosatsu/NormalizationTest.scala

```diff
@@ -0,0 +1,242 @@
+package org.bykn.bosatsu
+
+import org.scalatest.FunSuite
+import java.math.BigInteger
+import cats.data.NonEmptyList
+
+class NormalizationTest extends FunSuite {
+  import TestUtils._
+  import NormalExpression._
+  import Lit._
+  import Normalization._
+  import NormalPattern.{PositionalStruct, Var, ListPat}
+
+  test("Literal") {
+      normalTagTest(
+        List("""
+package NormTest/String
+
+main = "aa"
+"""
+        ), "NormTest/String", NormalExpressionTag(Literal(Str("aa")), Set())
+      )
+
+      normalTagTest(
+        List("""
+package NormTest/Int
+
+main = 22
+"""
+        ), "NormTest/Int", NormalExpressionTag(Literal(Integer(BigInteger.valueOf(22))), Set())
+      )
+
+      normalTagTest(
+        List("""
+package NormTest/List
+
+main = ["aa"]
+"""
+        ), "NormTest/List", NormalExpressionTag(
+          Struct(1,List(Literal(Str("aa")), Struct(0,List()))),
+          Set(
+            Lambda(Lambda(Struct(1,List(LambdaVar(1), LambdaVar(0))))),
+            Literal(Str("aa")),
+            Lambda(Struct(1,List(Literal(Str("aa")), LambdaVar(0)))),
+            Struct(0,List())
+          )
+        )
+      )
+  }
+  test("recurse") {
+    normalExpressionTest(
+      List("""
+package Recur/Some
+
+def foo(x):
+  recur x:
+    []: ["a","b","c"]
+    [h, *t]: NonEmptyList("zero", foo(t))
+
+out = foo
+"""
+      ), "Recur/Some", Recursion(Lambda(Lambda(Match(LambdaVar(0),NonEmptyList.fromList(List(
+        (ListPat(List()),Struct(1,List(Literal(Str("a")), Struct(1,List(Literal(Str("b")), Struct(1,List(Literal(Str("c")), Struct(0,List())))))))),
+        (ListPat(List(Right(Var(1)), Left(Some(0)))),Lambda(Lambda(Struct(1,List(Literal(Str("zero")), App(LambdaVar(3),LambdaVar(0)))))))
+      )).get))))
+    )
+  }
+  test("Lambda") {
+    normalTagTest(
+      List("""
+package Lambda/Identity
+
+out = \x -> x
+"""
+      ), "Lambda/Identity", NormalExpressionTag(
+        Lambda(LambdaVar(0)), Set(LambdaVar(0))
+      )
+    )
+    normalTagTest(
+      List("""
+package Lambda/Always
+
+out = \x -> \y -> x
+"""
+      ), "Lambda/Always", NormalExpressionTag(
+        Lambda(Lambda(LambdaVar(1))), Set(Lambda(LambdaVar(1)), LambdaVar(1))
```

Comment:
maybe make a function to build these:
```
def mkTag(ne: NormalExpression) = NormalExpressionTag(ne, Set(ne))
```
so it is clearer the part we are testing?

Ahh, I think this relates to my misunderstanding of the meaning of the Set. So, the right hand side is the set of all strict sub-expressions, so, which I guess we hope to form a lattice. Is that right?

Reply:
Totally happy to have a video chat

They are supposed to be a flat set of all descendants, so perhaps `children` is a bad name. Not that a NormalExpression does not have `children` themselves. But `NormalExpressionTag` does and that's on the tag to these `TypedExpr`.

I could actually remove these sets for now. I found them helpful in the past but they also seem to be a sledge hammer to a problem that might have another solution.

What they solve is in caching if you have
```
def bar(x):
  x.add(2)

def foo(x):
  bar(x)
```
Then you have an `AnnotatedLambda` for `foo` and for `bar`. After normalizations both of these expressions have the same `NormalExpression`. But it is important to note that `NormalExpressions` are just descriptors of the final outcome, the `TypedExpr`s themselves do not undergo normalization. So what happens if we add some caching to the `Evaluation` code. Both the `TypedExpr` for `foo` and `bar` have the same `NormalExpression` so if we use this as a key only one of them can be in the cache.

But the evaluation for a `AnnotatedLambda` is just a `Fn` that calls back into the evaluator with an `Env` and a `TypedExpr` in the closure.

What happened with my implementation is that `foo` would go into the cache, when it was called it would come to a call to `bar`, which it would look up in the cache, and find `foo` again.

So my hammer was to introduce these `children` and if a `TypedExpr` had its `ne` as a member of its `children` we would not write it to the cache

Reply:
Oh... I think I realize a way I'm still being unclear in my description. "They are supposed to be a flat set of all descendants" means it is the set of `NormalExpression`s for all the strict-sub `TypedExpr`s of the `TypedExpr` being tagged